app/src/main/java/org/sagebase/crf/step/heartrate/BpmRecorder.java

```diff
@@ -126,87 +181,105 @@ public void calculateBpm(HeartBeatSample heartBeatSample) {
         
         private final BpmRecorder.BpmUpdateListener mBpmUpdateListener;
         private final BpmRecorder.IntelligentStartUpdateListener mIntelligentStartListener;
-        
+        private final BpmRecorder.PressureListener mPressureListener;
+        private final BpmRecorder.CameraCoveredListener mCameraListener;
+        private final BpmRecorder.AbnormalHRListener mAbnormalListener;
+        private final BpmRecorder.DeclineHRListener mDeclineListener;
+
         private final Handler mainHandler = new Handler(Looper.getMainLooper());
         
         private final BpmCalculator bpmCalculator;
         
         public HeartBeatJsonWriter(BpmUpdateListener
                                            mBpmUpdateListener, IntelligentStartUpdateListener
                                            mIntelligentStartListener,
-                                   String identifier, Step step, File outputDirectory) {
+                                   CameraCoveredListener mCameraListener,
+                                   PressureListener mPressureListener, AbnormalHRListener mAbnormalListener,
+                                   DeclineHRListener mDeclineListener, String identifier,
+                                   Step step, File outputDirectory) {
             super(identifier, step, outputDirectory);
             
             this.mBpmUpdateListener = mBpmUpdateListener;
             this.mIntelligentStartListener = mIntelligentStartListener;
+            this.mPressureListener = mPressureListener;
+            this.mCameraListener = mCameraListener;
             this.bpmCalculator = new BpmCalculator();
+            this.mAbnormalListener = mAbnormalListener;
+            this.mDeclineListener = mDeclineListener;
         }
 
         private int sampleCount = 0;
         private double timestampReference = -1;
-        
+
         @AnyThread
         @Override
         public void onHeartRateSampleDetected(HeartBeatSample sample) {
-            bpmCalculator.calculateBpm(sample);
+            if(!sample.isCoveringLens()) {
+                mainHandler.post(() ->
+                        mCameraListener.cameraUpdate(new
+                                BpmRecorder.CameraCoveredListener.CameraCoveredHolder(false)));
 
-            // syoung 11/19/2018 Debug code added to get the sampling rate.
-            if (timestampReference == -1) {
-                timestampReference = sample.timestamp;
-            }
-            else if (sample.timestamp - timestampReference >= 1.0) {
-                LOG.debug("preprocessed sample count:{}", sampleCount);
-                timestampReference = sample.timestamp;
-                sampleCount = 0;
+                mIntelligentStartCounter = 0;
```

Comment:
Is this going to reset the measurement every time the lens is no longer covered?  For the stair step test I don't think we can reset the measurement.